### PR TITLE
drivers/gpio: Add support for GRLIB GRGPIO2

### PR DESCRIPTION
**Pull request by ghost**
This adds support for the GRLIB GRGPIO2 controller used in
LEON (SPARC) and NOEL-V (RISC-V) systems.

Also adds the GPIO controller description to the GR716A
MCU DTS and makes the GPIO option available in the kernel
configuration.

### CODE REVIEW

**Comment by galak** [comment]
Marking this DNM to see if we make progress on the new pinctrl API and thus this may require changes.

File: drivers/gpio/gpio_grgpio2.c
Comment (mnkp) [inline drivers/gpio/gpio_grgpio2.c:55]: We typically keep init, irq handler functions towards the bottom of the file. Just above the struct defining the API. Let's move it there.

Comment (mnkp) [inline drivers/gpio/gpio_grgpio2.c:55]: It still happens that people need to read disassembly code. A short, popular names like `init` make it difficult. Let's prefix the function, e.g. `grgpio2_init`.

File: soc/sparc/gr716a/Kconfig.defconfig
Comment (mnkp) [inline soc/sparc/gr716a/Kconfig.defconfig:22]: We generally prefer `depends on` rather than `if` #22666. Let's write it as
```
config GPIO_GRGPIO2
	default y
	depends on GPIO
```

File: drivers/gpio/gpio_grgpio2.c
Comment (mnkp) [inline drivers/gpio/gpio_grgpio2.c:86]: It looks like the driver doesn't support pull up / pull down, single-ended mode (open drain or open source). The function should return `-ENOTSUP` if relevant flags are passed.

**Review by mnkp — COMMENTED** [review]
LGTM. @martin-aberg Could you please test the driver with the following testcases
- tests/drivers/gpio/gpio_api_1pin
- tests/drivers/gpio/gpio_basic_api

Do you happen to have link to reference manual where GRGPIO2 module is described? From the point of view of software rather than hardware engineer.

File: drivers/gpio/gpio_grgpio2.c
Comment (martin-aberg) [reply drivers/gpio/gpio_grgpio2.c]: Yes the controller can read the input when configured as output. The driver has now been updated to support this. Thanks for the suggestion.

File: drivers/gpio/gpio_grgpio2.c
Comment (martin-aberg) [reply drivers/gpio/gpio_grgpio2.c]: Done

Comment (martin-aberg) [reply drivers/gpio/gpio_grgpio2.c]: Done

File: soc/sparc/gr716a/Kconfig.defconfig
Comment (martin-aberg) [reply soc/sparc/gr716a/Kconfig.defconfig]: Done. Thanks!

File: drivers/gpio/gpio_grgpio2.c
Comment (martin-aberg) [reply drivers/gpio/gpio_grgpio2.c:58]: OK! Done

**Comment by martin-aberg** [comment]
The reference manual for the GPIO controller can be found here:
  https://www.gaisler.com/doc/gr716/gr716-ds-um.pdf#G31.966345

**Comment by martin-aberg** [comment]
> Marking this DNM to see if we make progress on the new pinctrl API and thus this may require changes.

@galak, this is a GPIO driver. The pinctrl driver is #36326.

File: drivers/gpio/gpio_grgpio2.c
Comment (mnkp) [inline drivers/gpio/gpio_grgpio2.c:85]: If you happen to update this PR for some other reason you could add one empty line between variable definition block and the function body (even if the function is small).

**Review by mnkp — COMMENTED** [review]
LGTM. Since the driver supports now in/out pin configuration it should be enough to test it with `tests/drivers/gpio/gpio_api_1pin` testcase. Could you confirm that it is passing?

**Review by mbolivar-nordic — DISMISSED** [review]
DT usage seems fine

**Comment by galak** [comment]
Removed DNM as I don't see any reason we need to hold up GPIO driver once code is approved.

File: dts/bindings/gpio/gaisler,grgpio.yaml
Comment (galak) [inline dts/bindings/gpio/gaisler,grgpio.yaml:2]: Can we add a reference to the doc link you provided here.

File: dts/sparc/gr716a.dtsi
Comment (galak) [inline dts/sparc/gr716a.dtsi:109]: comment seems confusing compared to the value.

File: dts/sparc/gr716a.dtsi
Comment (galak) [inline dts/sparc/gr716a.dtsi:97]: Can you expand the comment, doesn't seem to convey much about what 17-20 means

**Review by galak — DISMISSED** [review]
LGTM a few minor bits here and there.

File: drivers/gpio/gpio_grgpio2.c
Comment (martin-aberg) [reply drivers/gpio/gpio_grgpio2.c:86]: Done

File: dts/bindings/gpio/gaisler,grgpio.yaml
Comment (martin-aberg) [reply dts/bindings/gpio/gaisler,grgpio.yaml:6]: It is added now. Thanks for the suggestion.

File: dts/sparc/gr716a.dtsi
Comment (martin-aberg) [reply dts/sparc/gr716a.dtsi:107]: Agreed. I have removed the comments. (The numbers are related to the interrupt controller priority/remap functionality which is currently not configured in Zephyr.)

**Comment by martin-aberg** [comment]
> 
> LGTM. Since the driver supports now in/out pin configuration it should be enough to test it with `tests/drivers/gpio/gpio_api_1pin` testcase. Could you confirm that it is passing?

Yes. I have run `tests/drivers/gpio/gpio_api_1pin`. All subtests pass except for the level interrupt tests (`test_gpio_int_level_*`).

Level interrupts are supported by the GPIO controller with a dedicated register bit which is activated with the `mode` `GPIO_INT_MODE_LEVEL`. However the LEON interrupt controller is designed for pulse interrupts. The consequence is that when a peripheral presents a continuous high level (rather than pulse) to the interrupt controller as in this case, the ISR will trig twice. I verified that this happened in the test by printing the ISR callback count and expected count.

So the level interrupt feature (`GPIO_INT_MODE_LEVEL`) is in the driver in case of custom use cases, or for use with an alternative interrupt controller.  Please let me know if you would like to see that being removed instead.

The edge trigger interrupt tests, and the other tests passed as expected.

File: drivers/gpio/gpio_grgpio2.c
Comment (mnkp) [inline drivers/gpio/gpio_grgpio2.c:171]: If we cleared here respective bit in `regs->iflag` register would that solve the issue with level interrupts?

File: drivers/gpio/gpio_grgpio2.c
Comment (mnkp) [inline drivers/gpio/gpio_grgpio2.c:237]: I believe this check should be an assert, or we can remove it completely. If the driver is bug free it's not possible to have an interrupt without an interrupt source?

File: drivers/gpio/gpio_grgpio2.c
Comment (mnkp) [reply drivers/gpio/gpio_grgpio2.c:171]: Actually, it would be more appropriate to clear `regs->iflag` shortly before enabling the interrupt and not after disabling it.

File: drivers/gpio/gpio_grgpio2.c
Comment (martin-aberg) [reply drivers/gpio/gpio_grgpio2.c:240]: It is possible that the ISR gets called when `regs->iflag == 0`.

The `iflag` is a status indicator local to the GPIO controller. A bit gets set to `1` in `iflag` when the corresponding I/O bit causes an interrupt condition. That condition is also forwarded to the interrupt controller as a request to set one of 31 pending bits.

A bit in `iflag` can be cleared by the user by writing `1`.  However, that write only affects the GPIO status indicator and does not acknowledge pending bits in the interrupt controller. The driver can clear bits in `iflag` which correspond to interrupts which are still pending in the interrupt controller. In that scenario, the ISR may get called after the driver has already cleared the `iflag` indication.

Acknowledgement of a pending interrupt in the LEON interrupt controller is performed automatically by hardware, at the same moment as the CPU starts executing the first instruction of the interrupt exception handler.

File: drivers/gpio/gpio_grgpio2.c
Comment (martin-aberg) [reply drivers/gpio/gpio_grgpio2.c:171]: Agreed. The respective bit in `iflag` should be cleared shortly before enabling the interrupt. It has been added in the latest push. The flag gets cleared also when disabling interrupt for a pin, and that solved the issue with level interrupts in the test case. Thanks!

All subtests in `gpio_api_1pin_test` now pass.

**Review by mnkp — DISMISSED** [review]

**Comment by martin-aberg** [comment]
@galak All review comments should have been addressed now. Do you think we can move this forward?

**Comment by mnkp** [comment]
@martin-aberg could you rebase this PR on the latest master.Thanks!

**Comment by mnkp** [comment]
@martin-aberg some checks are still failing. Could you have a look.

**Pull request closed without merging**
